oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,522 @@
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.json.webtoken.JsonWebSignature;
+import com.google.api.client.json.webtoken.JsonWebToken;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.GeneralSecurityException;
+import java.security.PrivateKey;
+import java.util.Date;
+import java.util.Map;
+import java.util.Objects;
+
+public class GdchCredentials extends GoogleCredentials {
+  static final String SUPPORTED_FORMAT_VERSION = "1";
+  private static final String PARSE_ERROR_PREFIX = "Error parsing token refresh response. ";
+  private static final String GDCH_ISSUER_SUBJECT_VALUE_PREFIX = "system:serviceaccount:%s:%s";
+  private static final int DEFAULT_LIFETIME_IN_SECONDS = 3600;
+
+  private final PrivateKey privateKey;
+  private final String privateKeyId;
+  private final String projectId;
+  private final String serviceIdentityName;
+  private final URI tokenServerUri;
+  private final URI apiAudience;
+  private final int lifetime;
+  private final String transportFactoryClassName;
+  private final String caCertPath;
+  private transient HttpTransportFactory transportFactory;
+
+  /**
+   * Internal constructor.
+   *
+   * @param builder A builder for {@link GdchCredentials} See {@link GdchCredentials.Builder}.
+   */
+  @VisibleForTesting
+  GdchCredentials(GdchCredentials.Builder builder) {
+    this.projectId = Preconditions.checkNotNull(builder.projectId);
+    this.privateKeyId = Preconditions.checkNotNull(builder.privateKeyId);
+    this.privateKey = Preconditions.checkNotNull(builder.privateKey);
+    this.serviceIdentityName = Preconditions.checkNotNull(builder.serviceIdentityName);
+    this.tokenServerUri = Preconditions.checkNotNull(builder.tokenServerUri);
+    this.transportFactory = Preconditions.checkNotNull(builder.transportFactory);
+    this.transportFactoryClassName = this.transportFactory.getClass().getName();
+    this.caCertPath = builder.caCertPath;
+    this.apiAudience = builder.apiAudience;
+    this.lifetime = builder.lifetime;
+  }
+
+  /**
+   * Create GDCH service account credentials defined by JSON.
+   *
+   * @param json a map from the JSON representing the credentials.
+   * @return the GDCH service account credentials defined by the JSON.
+   * @throws IOException if the credential cannot be created from the JSON.
+   */
+  static GdchCredentials fromJson(Map<String, Object> json) throws IOException {
+    String caCertPath = validateField((String) json.get("ca_cert_path"), "ca_cert_path");
```

Comment:
It seems the `validateField` will throw exceptions if `ca_cert_path` is missing. `ca_cert_path` is not a required field so if it is missing we just use the regular transportFactory. I think you can change the GDCHTransportFactory implementation, if ca cert path missing just return a regular one.

Reply:
Thanks! I changed `ca_cert_path` to optional and TransportFactoryForGdch implementation. Also, modified related test cases.